// @betty-blocks/cli/build/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var InteractionOptionType;
(function (InteractionOptionType) {
    InteractionOptionType["Boolean"] = "Boolean";
    InteractionOptionType["Number"] = "Number";
    InteractionOptionType["String"] = "String";
    InteractionOptionType["Event"] = "Event";
    InteractionOptionType["Void"] = "Void";
    InteractionOptionType["Page"] = "Page";
})(InteractionOptionType = exports.InteractionOptionType || (exports.InteractionOptionType = {}));
// TODO: Add support
var InteractionOptionTypeToDo;
(function (InteractionOptionTypeToDo) {
    InteractionOptionTypeToDo["Color"] = "Color";
    InteractionOptionTypeToDo["Endpoint"] = "Endpoint";
    InteractionOptionTypeToDo["Filter"] = "Filter";
    InteractionOptionTypeToDo["Font"] = "Font";
    InteractionOptionTypeToDo["Properties"] = "Properties";
    InteractionOptionTypeToDo["Property"] = "Property";
    InteractionOptionTypeToDo["Size"] = "Size";
    InteractionOptionTypeToDo["Unit"] = "Unit";
})(InteractionOptionTypeToDo = exports.InteractionOptionTypeToDo || (exports.InteractionOptionTypeToDo = {}));
var InteractionType;
(function (InteractionType) {
    InteractionType["Custom"] = "Custom";
    InteractionType["Global"] = "Global";
})(InteractionType = exports.InteractionType || (exports.InteractionType = {}));
