// @betty-blocks/cli/build/validations/prefab/component.js
"use strict";
/* eslint-disable no-use-before-define */
/* eslint-disable @typescript-eslint/no-use-before-define */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const joi_1 = __importDefault(require("joi"));
const validation_1 = require("../../utils/validation");
const componentOption_1 = require("./componentOption");
const componentSchema = joi_1.default.object({
    name: joi_1.default.string().required(),
    ref: joi_1.default.object({
        id: joi_1.default.string().required(),
    }),
    options: joi_1.default.array()
        .items(componentOption_1.optionSchema)
        .required(),
    descendants: joi_1.default.array()
        .items(joi_1.default.custom(validateComponent))
        .required(),
});
function validateComponent(component) {
    const { name, options } = component;
    const { error } = componentSchema.validate(component);
    validation_1.findDuplicates(options, 'option key', 'key');
    if (typeof error !== 'undefined') {
        const { message } = error;
        throw new Error(chalk_1.default.red(`\nBuild error in component ${name}: ${message}\n`));
    }
    return component;
}
exports.validateComponent = validateComponent;
