// @betty-blocks/cli/build/validations/prefab/interaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("joi"));
const constants_1 = require("../constants");
const parametersSchema = joi_1.default.when('type', {
    is: 'Global',
    then: joi_1.default.array()
        .items(joi_1.default.object({
        parameter: joi_1.default.string().required(),
        name: joi_1.default.string().required(),
        ref: joi_1.default.object({
            componentId: joi_1.default.string().required(),
        }).required(),
    }))
        .required(),
    otherwise: joi_1.default.forbidden(),
});
const targetOptionNameSchema = joi_1.default.when('type', {
    is: 'Global',
    then: joi_1.default.string(),
    otherwise: joi_1.default.forbidden(),
});
const targetComponentIdSchema = joi_1.default.when('type', {
    is: 'Custom',
    then: joi_1.default.string().required(),
});
exports.interactionSchema = joi_1.default.object({
    name: joi_1.default.string().required(),
    ref: joi_1.default.object({
        sourceComponentId: joi_1.default.string().required(),
        targetComponentId: targetComponentIdSchema,
    }).required(),
    targetOptionName: targetOptionNameSchema,
    sourceEvent: joi_1.default.string().required(),
    type: joi_1.default.string()
        .valid(...constants_1.INTERACTION_TYPE)
        .required(),
    parameters: parametersSchema,
});
