// @betty-blocks/cli/build/validations/prefab/variable.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("joi"));
const constants_1 = require("../constants");
exports.variableKindSchema = joi_1.default.when('kind', {
    switch: [
        {
            is: 'construct',
            then: joi_1.default.object({
                modelId: joi_1.default.string()
                    .allow('')
                    .required(),
                ref: joi_1.default.object({
                    customModelId: joi_1.default.string().required(),
                }).required(),
            }),
        },
        {
            is: 'object',
            then: joi_1.default.object({
                modelId: joi_1.default.string()
                    .allow('')
                    .required(),
            }),
        },
    ],
    otherwise: joi_1.default.forbidden(),
});
exports.variableSchema = joi_1.default.object({
    name: joi_1.default.string()
        .allow('')
        .required(),
    kind: joi_1.default.string()
        .valid(...constants_1.VARIABLE_KIND)
        .required(),
    pageId: joi_1.default.string().allow(''),
    ref: joi_1.default.object({
        id: joi_1.default.string().required(),
        endpointId: joi_1.default.string(),
        actionId: joi_1.default.string(),
    }).required(),
    options: exports.variableKindSchema,
});
