// @betty-blocks/cli/build/bb-components-generate.js
"use strict";
/* npm dependencies */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const commander_1 = __importDefault(require("commander"));
const chalk_1 = __importDefault(require("chalk"));
const fs_extra_1 = require("fs-extra");
const checkUpdateAvailable_1 = require("./utils/checkUpdateAvailable");
/* process arguments */
commander_1.default
    .usage('[name]')
    .name('bb components generate')
    .parse(process.argv);
const { args } = commander_1.default;
if (args.length === 0) {
    commander_1.default.help();
}
const name = args[0];
(() => __awaiter(void 0, void 0, void 0, function* () {
    yield checkUpdateAvailable_1.checkUpdateAvailableCLI();
    if (name.includes(' ')) {
        throw new Error(chalk_1.default.red(`\nName cannot contain spaces\n`));
    }
    if (yield fs_extra_1.pathExists(`src/prefabs/${name}.js`)) {
        throw new Error(chalk_1.default.red(`\nPrefab ${name} already exists\n`));
    }
    if (yield fs_extra_1.pathExists(`src/components/${name}.js`)) {
        throw new Error(chalk_1.default.red(`\nComponent ${name} already exists\n`));
    }
    const prefab = `
(() => ({
  name: '${name}',
  icon: 'TitleIcon',
  category: 'CONTENT',
  structure: [
    {
      name: '${name}',
      options: [],
      descendants: [],
    },
  ],
}))();
  `;
    const component = `
(() => ({
  name: '${name}',
  type: 'ROW',
  allowedTypes: [],
  orientation: 'HORIZONTAL',
  jsx: <div className={classes.root}>Hello World</div>,
  styles: () => () => ({
    root: {},
  }),
}))();
  `;
    yield Promise.all([
        fs_extra_1.outputFile(`src/prefabs/${name}.js`, prefab.trim()),
        fs_extra_1.outputFile(`src/components/${name}.js`, component.trim()),
        console.log(chalk_1.default.green('The component has been generated')),
    ]);
}))();
