// @betty-blocks/cli/build/validations/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_ACTIONS = 5;
exports.MAX_VARIABLES = 20;
exports.MAX_INTERACTIONS = 60;
exports.MAX_ACTION_EVENTS = 50;
exports.ICONS = [
    'AccordionIcon',
    'AccordionItemIcon',
    'AlertIcon',
    'AutoCompleteIcon',
    'AvatarIcon',
    'BreadcrumbIcon',
    'BreadcrumbItemIcon',
    'ButtonGroupIcon',
    'ButtonIcon',
    'CardIcon',
    'CardActionsIcon',
    'CardContentIcon',
    'CardHeaderIcon',
    'CardMediaIcon',
    'CheckboxIcon',
    'CheckboxGroupIcon',
    'ChipIcon',
    'Column2Icon',
    'Column3Icon',
    'ColumnIcon',
    'ConditionalIcon',
    'ContainerIcon',
    'CreateFormIcon',
    'DataContainer',
    'DataList',
    'DataTable',
    'DataTableBody',
    'DataTableColumn',
    'DataTableHead',
    'DataTableRow',
    'DatePickerIcon',
    'DateTimePickerIcon',
    'DecimalInputIcon',
    'DefinitionListIcon',
    'DeleteRecordIcon',
    'DetailViewIcon',
    'DialogIcon',
    'DrawerIcon',
    'DynamicFormIcon',
    'DynamicTableIcon',
    'DynamicTilesIcon',
    'EmailInputIcon',
    'FileInputIcon',
    'FormIcon',
    'GridIcon',
    'HiddenInputIcon',
    'HorizontalRuleIcon',
    'HtmlIcon',
    'IbanInputIcon',
    'IconIcon',
    'ImageCarouselIcon',
    'ImageIcon',
    'ImageInputIcon',
    'IncludeIcon',
    'LabelIcon',
    'Layout1Icon',
    'Layout2Icon',
    'Layout363Icon',
    'Layout444Icon',
    'Layout48Icon',
    'Layout66Icon',
    'Layout3333Icon',
    'Layout84Icon',
    'LogoutIcon',
    'ListItemIcon',
    'ListWithDataIcon',
    'LoginFormIcon',
    'MenuIcon',
    'MenuItemIcon',
    'MultiLineIcon',
    'MultiSelectIcon',
    'NavbarIcon',
    'NavItemIcon',
    'NavSidebarIcon',
    'NumberInputIcon',
    'OpenPageIcon',
    'OrderedListIcon',
    'PanelIcon',
    'PaperIcon',
    'ParagraphIcon',
    'PasswordInputIcon',
    'PhoneInputIcon',
    'PriceInputIcon',
    'ProgressBarIcon',
    'RadioButtonIcon',
    'RatingIcon',
    'RowIcon',
    'RowColumnIcon',
    'SelectIcon',
    'SidebarLeftIcon',
    'SliderComponentIcon',
    'SnackbarIcon',
    'StepIcon',
    'StepperIcon',
    'SubmitButtonIcon',
    'SwitcherIcon',
    'TabIcon',
    'TabsIcon',
    'Table',
    'TextareaIcon',
    'TextInputIcon',
    'TimePickerIcon',
    'TitleIcon',
    'UnorderedListIcon',
    'UpdateFormIcon',
    'UrlInputIcon',
];
exports.OPTIONS = [
    'ACTION',
    'ACTION_INPUT',
    'ACTION_INPUT_OBJECTS',
    'ACTION_PROPERTIES',
    'AUTHENTICATION_PROFILE',
    'BUTTONGROUP',
    'COLOR',
    'CUSTOM',
    'CUSTOM_MODEL_ATTRIBUTE',
    'DROPDOWN',
    'ENDPOINT',
    'FILTER',
    'FONT',
    'FORM_DATA',
    'MODEL',
    'MODELS',
    'NUMBER',
    'PROPERTY',
    'PROPERTIES',
    'SIZE',
    'SIZES',
    'TEXT',
    'TOGGLE',
    'UNIT',
    'VARIABLE',
];
exports.STYLE_REFERENCE_TYPES = ['BUTTON', 'TEXT'];
exports.CONDITION_TYPE = ['HIDE', 'SHOW'];
exports.COMPARATORS = ['EQ'];
exports.MODAL_TYPE = ['MODEL_AND_PROPERTIES'];
exports.CONFIGURATION_AS = [
    'BUTTONGROUP',
    'DROPDOWN',
    'MULTILINE',
    'UNIT',
    'VISIBILITY',
];
exports.INTERACTION_TYPE = ['Global', 'Custom'];
exports.EVENT_KIND_NEW_RUNTIME = 'authenticate_user';
exports.EVENT_KIND = [
    'action',
    'assign',
    'auto_increment_generate',
    'auto_increment_set',
    'condition',
    'create',
    'create_betty_user',
    'delete',
    'export',
    'expression',
    'group',
    'http_request',
    'import',
    'login_web_user',
    'logout_web_user',
    'loop',
    'pdf_generate',
    'pdf_merge',
    'redirect_web_page',
    'render_web_template',
    'send_mail',
    'sftp_download',
    'sftp_list',
    'sftp_upload',
    'update',
    'zip',
];
exports.VARIABLE_KIND = ['construct', 'object', 'string', 'integer'];
