// @betty-blocks/cli/build/validations/component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const joi_1 = __importDefault(require("joi"));
const validation_1 = require("../utils/validation");
const constants_1 = require("./constants");
const schema = joi_1.default.object({
    name: joi_1.default.string().required(),
    icon: joi_1.default.string(),
    category: joi_1.default.string(),
    type: joi_1.default.string().required(),
    allowedTypes: joi_1.default.array()
        .items(joi_1.default.string())
        .required(),
    orientation: joi_1.default.string().required(),
    functions: joi_1.default.array().items(joi_1.default.string()),
    triggers: joi_1.default.array().items(joi_1.default.string()),
    interactions: joi_1.default.object(),
    jsx: joi_1.default.any().required(),
    styles: joi_1.default.any().required(),
    styleType: joi_1.default.string().valid(...constants_1.STYLE_REFERENCE_TYPES),
});
const validate = (component) => {
    const { error } = schema.validate(component);
    if (typeof error !== 'undefined') {
        throw new Error(chalk_1.default.red(`\nProperty: ${error.message} at component: ${component.name}\n`));
    }
};
exports.default = (components) => {
    components.forEach((component) => {
        validate(component);
    });
    validation_1.findDuplicates(components, 'component', 'name');
};
